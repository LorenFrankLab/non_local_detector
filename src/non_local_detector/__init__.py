from non_local_detector.models import (
    ContFragClusterlessClassifier,  # noqa
    ContFragSortedSpikesClassifier,  # noqa
    MultiEnvironmentSortedSpikesClassifier,  # noqa
    MultiEnvironmentClusterlessClassifier,  # noqa
    NonLocalClusterlessDetector,  # noqa
    NonLocalSortedSpikesDetector,  # noqa
    ClusterlessDecoder,  # noqa
    SortedSpikesDecoder,  # noqa
)

try:
    from ._version import __version__
except ImportError:
    pass
