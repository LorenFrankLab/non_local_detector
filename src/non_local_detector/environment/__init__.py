from non_local_detector.environment.alignment import (
    get_2d_rotation_matrix,
    map_probabilities_to_nearest_target_bin,
)
from non_local_detector.environment.environment import Environment
from non_local_detector.environment.layout.factories import (
    get_layout_parameters,
    list_available_layouts,
)

__all__ = [
    "get_2d_rotation_matrix",
    "map_probabilities_to_nearest_target_bin",
    "Environment",
    "get_layout_parameters",
    "list_available_layouts",
]
