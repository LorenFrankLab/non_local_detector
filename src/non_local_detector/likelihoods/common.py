from dataclasses import dataclass
from typing import Optional

import jax
import jax.numpy as jnp
import numpy as np
import scipy.interpolate
from track_linearization import get_linearized_position

from non_local_detector.environment import Environment

EPS = 1e-15
LOG_EPS = np.log(EPS)


def get_position_at_time(
    time: jnp.ndarray,
    position: jnp.ndarray,
    spike_times: jnp.ndarray,
    env: Optional[Environment] = None,
):
    position_at_spike_times = scipy.interpolate.interpn(
        (time,), position, spike_times, bounds_error=False, fill_value=None
    )
    if env is not None and env.track_graph is not None:
        if position_at_spike_times.shape[0] > 0:
            position_at_spike_times = get_linearized_position(
                position_at_spike_times,
                env.track_graph,
                edge_order=env.edge_order,
                edge_spacing=env.edge_spacing,
            ).linear_position.to_numpy()[:, None]
        else:
            position_at_spike_times = jnp.array([])[:, None]

    return position_at_spike_times


@jax.jit
def gaussian_pdf(x: jnp.ndarray, mean: jnp.ndarray, sigma: jnp.ndarray) -> jnp.ndarray:
    """Compute the value of a Gaussian probability density function at x with
    given mean and sigma."""
    return jnp.exp(-0.5 * ((x - mean) / sigma) ** 2) / (sigma * jnp.sqrt(2.0 * jnp.pi))


def kde(
    eval_points: jnp.ndarray, samples: jnp.ndarray, std: jnp.ndarray
) -> jnp.ndarray:
    distance = jnp.ones((samples.shape[0], eval_points.shape[0]))

    for dim_eval_points, dim_samples, dim_std in zip(eval_points.T, samples.T, std):
        distance *= gaussian_pdf(
            jnp.expand_dims(dim_eval_points, axis=0),
            jnp.expand_dims(dim_samples, axis=1),
            dim_std,
        )
    return jnp.mean(distance, axis=0).squeeze()


def block_kde(
    eval_points: jnp.ndarray,
    samples: jnp.ndarray,
    std: jnp.ndarray,
    block_size: int = 100,
) -> jnp.ndarray:
    n_eval_points = eval_points.shape[0]
    density = jnp.zeros((n_eval_points,))
    for start_ind in range(0, n_eval_points, block_size):
        block_inds = slice(start_ind, start_ind + block_size)
        density = jax.lax.dynamic_update_slice(
            density,
            kde(eval_points[block_inds], samples, std).squeeze(),
            (start_ind,),
        )

    return density


def block_kde(
    eval_points: jnp.ndarray,
    samples: jnp.ndarray,
    std: jnp.ndarray,
    block_size: int = 100,
) -> jnp.ndarray:
    n_eval_points = eval_points.shape[0]
    density = jnp.zeros((n_eval_points,))
    for start_ind in range(0, n_eval_points, block_size):
        block_inds = slice(start_ind, start_ind + block_size)
        density = jax.lax.dynamic_update_slice(
            density,
            kde(eval_points[block_inds], samples, std).squeeze(),
            (start_ind,),
        )

    return density


@dataclass
class KDEModel:
    std: jnp.ndarray
    block_size: Optional[int] = None

    def fit(self, samples: jnp.ndarray):
        samples = jnp.asarray(samples)
        if samples.ndim == 1:
            samples = jnp.expand_dims(samples, axis=1)
        self.samples_ = samples

        return self

    def predict(self, eval_points: jnp.ndarray):
        if eval_points.ndim == 1:
            eval_points = jnp.expand_dims(eval_points, axis=1)
        std = (
            jnp.array([self.std] * eval_points.shape[1])
            if isinstance(self.std, (int, float))
            else self.std
        )
        block_size = (
            eval_points.shape[0] if self.block_size is None else self.block_size
        )

        return block_kde(eval_points, self.samples_, std, block_size)


def get_spikecount_per_time_bin(spike_times, time):
    spike_times = spike_times[
        np.logical_and(spike_times >= time[0], spike_times <= time[-1])
    ]
    return np.bincount(
        np.digitize(spike_times, time[1:-1]),
        minlength=time.shape[0],
    )
