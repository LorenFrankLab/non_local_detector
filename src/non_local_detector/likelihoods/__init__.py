from non_local_detector.likelihoods.clusterless_kde import (
    fit_clusterless_kde_encoding_model,
    predict_clusterless_kde_log_likelihood,
)
from non_local_detector.likelihoods.no_spike import (
    predict_no_spike_log_likelihood,
)  # noqa
from non_local_detector.likelihoods.sorted_spikes_glm import (
    fit_sorted_spikes_glm_encoding_model,
    predict_sorted_spikes_glm_log_likelihood,
)
from non_local_detector.likelihoods.sorted_spikes_kde import (
    fit_sorted_spikes_kde_encoding_model,
    predict_sorted_spikes_kde_log_likelihood,
)

_SORTED_SPIKES_ALGORITHMS = {
    "sorted_spikes_glm": (
        fit_sorted_spikes_glm_encoding_model,
        predict_sorted_spikes_glm_log_likelihood,
    ),
    "sorted_spikes_kde": (
        fit_sorted_spikes_kde_encoding_model,
        predict_sorted_spikes_kde_log_likelihood,
    ),
}
_CLUSTERLESS_ALGORITHMS = {
    "clusterless_kde": (
        fit_clusterless_kde_encoding_model,
        predict_clusterless_kde_log_likelihood,
    ),
}
