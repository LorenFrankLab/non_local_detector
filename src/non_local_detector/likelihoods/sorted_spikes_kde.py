import jax
import jax.numpy as jnp
import numpy as np
from tqdm.autonotebook import tqdm
from track_linearization import get_linearized_position

from non_local_detector.environment import Environment
from non_local_detector.likelihoods.common import EPS, KDEModel, get_position_at_time


def fit_sorted_spikes_kde_encoding_model(
    position_time: jnp.ndarray,
    position: jnp.ndarray,
    spike_times: list[jnp.ndarray],
    environment: Environment,
    sampling_frequency: int = 500,
    position_std: float = 6.0,
    block_size: int = 100,
    disable_progress_bar: bool = False,
):
    position = position if position.ndim > 1 else jnp.expand_dims(position, axis=1)
    if isinstance(position_std, (int, float)):
        if environment.track_graph is not None and position.shape[1] > 1:
            position_std = jnp.array([position_std])
        else:
            position_std = jnp.array([position_std] * position.shape[1])

    is_track_interior = environment.is_track_interior_.ravel()
    interior_place_bin_centers = environment.place_bin_centers_[is_track_interior]

    if environment.track_graph is not None and position.shape[1] > 1:
        # convert to 1D
        position1D = get_linearized_position(
            position,
            environment.track_graph,
            edge_order=environment.edge_order,
            edge_spacing=environment.edge_spacing,
        ).linear_position.to_numpy()[:, None]
        occupancy_model = KDEModel(std=position_std, block_size=block_size).fit(
            position1D
        )
    else:
        occupancy_model = KDEModel(std=position_std, block_size=block_size).fit(
            position
        )

    occupancy = occupancy_model.predict(interior_place_bin_centers)

    mean_rates = []
    place_fields = []
    marginal_models = []

    n_time_bins = int((position_time[-1] - position_time[0]) * sampling_frequency)

    for neuron_spike_times in tqdm(
        spike_times,
        unit="cell",
        desc="Encoding models",
        disable=disable_progress_bar,
    ):
        neuron_spike_times = neuron_spike_times[
            jnp.logical_and(
                neuron_spike_times >= position_time[0],
                neuron_spike_times <= position_time[-1],
            )
        ]
        mean_rates.append(len(neuron_spike_times) / n_time_bins)
        neuron_marginal_model = KDEModel(std=position_std, block_size=block_size).fit(
            get_position_at_time(
                position_time, position, neuron_spike_times, environment
            )
        )
        marginal_models.append(neuron_marginal_model)
        marginal_density = neuron_marginal_model.predict(interior_place_bin_centers)
        marginal_density = jnp.where(jnp.isnan(marginal_density), 0.0, marginal_density)
        place_fields.append(
            jnp.zeros((is_track_interior.shape[0],))
            .at[is_track_interior]
            .set(
                jnp.clip(
                    mean_rates[-1]
                    * jnp.where(occupancy > 0.0, marginal_density / occupancy, EPS),
                    a_min=EPS,
                    a_max=None,
                )
            )
        )

    place_fields = jnp.stack(place_fields, axis=0)
    no_spike_part_log_likelihood = jnp.sum(place_fields, axis=0)

    return {
        "environment": environment,
        "marginal_models": marginal_models,
        "occupancy_model": occupancy_model,
        "occupancy": occupancy,
        "mean_rates": mean_rates,
        "place_fields": place_fields,
        "no_spike_part_log_likelihood": no_spike_part_log_likelihood,
        "is_track_interior": is_track_interior,
        "disable_progress_bar": disable_progress_bar,
    }


def predict_sorted_spikes_kde_log_likelihood(
    time: jnp.ndarray,
    position_time: jnp.ndarray,
    position: jnp.ndarray,
    spike_times: list[np.ndarray],
    environment: Environment,
    marginal_models: list[KDEModel],
    occupancy_model: KDEModel,
    occupancy: jnp.ndarray,
    mean_rates: jnp.ndarray,
    place_fields: jnp.ndarray,
    no_spike_part_log_likelihood: jnp.ndarray,
    is_track_interior: jnp.ndarray,
    disable_progress_bar: bool = False,
    is_local: bool = False,
):
    n_time = time.shape[0]
    if is_local:
        log_likelihood = jnp.zeros((n_time,))

        # Need to interpolate position
        interpolated_position = get_position_at_time(
            position_time, position, time, environment
        )
        occupancy = occupancy_model.predict(interpolated_position)

        for neuron_spike_times, neuron_marginal_model, neuron_mean_rate in zip(
            tqdm(
                spike_times,
                unit="cell",
                desc="Local Likelihood",
                disable=disable_progress_bar,
            ),
            marginal_models,
            mean_rates,
        ):
            neuron_spike_times = neuron_spike_times[
                np.logical_and(
                    neuron_spike_times >= time[0],
                    neuron_spike_times <= time[-1],
                )
            ]
            spike_count_per_time_bin = np.bincount(
                np.digitize(neuron_spike_times, time[1:-1]),
                minlength=time.shape[0],
            )
            marginal_density = neuron_marginal_model.predict(interpolated_position)
            marginal_density = jnp.where(
                jnp.isnan(marginal_density), 0.0, marginal_density
            )
            local_rate = neuron_mean_rate * jnp.where(
                occupancy > 0.0, marginal_density / occupancy, EPS
            )
            local_rate = jnp.clip(local_rate, a_min=EPS, a_max=None)
            log_likelihood += (
                jax.scipy.special.xlogy(spike_count_per_time_bin, local_rate)
                - local_rate
            )

        log_likelihood = jnp.expand_dims(log_likelihood, axis=1)
    else:
        n_interior_bins = is_track_interior.sum()
        log_likelihood = jnp.zeros((n_time, n_interior_bins))
        for neuron_spike_times, place_field in zip(
            tqdm(
                spike_times,
                unit="cell",
                desc="Non-Local Likelihood",
                disable=disable_progress_bar,
            ),
            place_fields,
        ):
            neuron_spike_times = neuron_spike_times[
                np.logical_and(
                    neuron_spike_times >= time[0],
                    neuron_spike_times <= time[-1],
                )
            ]
            spike_count_per_time_bin = np.bincount(
                np.digitize(neuron_spike_times, time[1:-1]),
                minlength=time.shape[0],
            )
            log_likelihood += jax.scipy.special.xlogy(
                np.expand_dims(spike_count_per_time_bin, axis=1),
                jnp.expand_dims(place_field[is_track_interior], axis=0),
            )

        log_likelihood -= no_spike_part_log_likelihood[is_track_interior]

    return log_likelihood
