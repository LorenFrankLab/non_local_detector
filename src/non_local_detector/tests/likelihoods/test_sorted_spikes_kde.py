import jax.numpy as jnp
import numpy as np
import pytest
from syrupy.assertion import SnapshotAssertion

from non_local_detector.environment import Environment
from non_local_detector.likelihoods.common import EPS, get_position_at_time
from non_local_detector.likelihoods.sorted_spikes_kde import (
    fit_sorted_spikes_kde_encoding_model,
    predict_sorted_spikes_kde_log_likelihood,
)


def test_fit_sorted_spikes_kde_encoding_model_minimal(simple_1d_environment):
    env = simple_1d_environment
    t = jnp.linspace(0.0, 10.0, 101)
    pos = jnp.linspace(0.0, 10.0, 101)[:, None]
    # two neurons with a couple spikes each inside [0,10]
    spikes = [jnp.array([2.0, 5.0, 5.1]), jnp.array([1.5, 7.2])]
    weights = jnp.ones_like(t)

    enc = fit_sorted_spikes_kde_encoding_model(
        position_time=t,
        position=pos,
        spike_times=spikes,
        environment=env,
        weights=weights,
        sampling_frequency=10,
        position_std=np.sqrt(1.0),
        block_size=16,
        disable_progress_bar=True,
    )

    # keys and shapes
    for key in (
        "environment",
        "marginal_models",
        "occupancy_model",
        "occupancy",
        "mean_rates",
        "place_fields",
        "no_spike_part_log_likelihood",
        "is_track_interior",
    ):
        assert key in enc
    occ = enc["occupancy"]
    pf = enc["place_fields"]
    assert occ.ndim == 1 and occ.size > 0
    assert jnp.all(jnp.isfinite(occ))
    assert pf.ndim == 2 and pf.shape[0] == len(spikes)
    assert jnp.all(pf > 0)


def test_predict_sorted_spikes_kde_log_likelihood_shapes_local_and_nonlocal(
    simple_1d_environment,
):
    env = simple_1d_environment
    t_pos = jnp.linspace(0.0, 10.0, 101)
    pos = jnp.linspace(0.0, 10.0, 101)[:, None]
    spikes = [jnp.array([2.0, 5.0, 5.1]), jnp.array([1.5, 7.2])]
    weights = jnp.ones_like(t_pos)

    enc = fit_sorted_spikes_kde_encoding_model(
        position_time=t_pos,
        position=pos,
        spike_times=spikes,
        environment=env,
        weights=weights,
        sampling_frequency=10,
        position_std=np.sqrt(1.0),
        block_size=16,
        disable_progress_bar=True,
    )

    # decoding over 5 time bins via 6 edges
    t_edges = jnp.linspace(0.0, 10.0, 6)

    # Non-local
    ll = predict_sorted_spikes_kde_log_likelihood(
        time=t_edges,
        position_time=t_pos,
        position=pos,
        spike_times=spikes,
        environment=env,
        marginal_models=enc["marginal_models"],
        occupancy_model=enc["occupancy_model"],
        occupancy=enc["occupancy"],
        mean_rates=jnp.asarray(enc["mean_rates"]),
        place_fields=enc["place_fields"],
        no_spike_part_log_likelihood=enc["no_spike_part_log_likelihood"],
        is_track_interior=enc["is_track_interior"],
        disable_progress_bar=True,
        is_local=False,
    )
    assert ll.shape[0] == t_edges.shape[0]
    assert ll.ndim == 2 and ll.shape[1] == int(enc["is_track_interior"].sum())
    assert jnp.all(jnp.isfinite(ll))

    # Local
    ll_local = predict_sorted_spikes_kde_log_likelihood(
        time=t_edges,
        position_time=t_pos,
        position=pos,
        spike_times=spikes,
        environment=env,
        marginal_models=enc["marginal_models"],
        occupancy_model=enc["occupancy_model"],
        occupancy=enc["occupancy"],
        mean_rates=jnp.asarray(enc["mean_rates"]),
        place_fields=enc["place_fields"],
        no_spike_part_log_likelihood=enc["no_spike_part_log_likelihood"],
        is_track_interior=enc["is_track_interior"],
        disable_progress_bar=True,
        is_local=True,
    )
    assert ll_local.shape == (t_edges.shape[0], 1)
    assert jnp.all(jnp.isfinite(ll_local))


def test_local_likelihood_zero_spikes_equals_negative_rate_sum(simple_1d_environment):
    env = simple_1d_environment
    t_pos = jnp.linspace(0.0, 10.0, 101)
    pos = jnp.linspace(0.0, 10.0, 101)[:, None]
    # Use some spikes to build encoding, but no spikes in decoding window
    spikes_enc = [jnp.array([2.0, 5.0, 5.1]), jnp.array([1.5, 7.2])]
    weights = jnp.ones_like(t_pos)

    enc = fit_sorted_spikes_kde_encoding_model(
        position_time=t_pos,
        position=pos,
        spike_times=spikes_enc,
        environment=env,
        weights=weights,
        sampling_frequency=10,
        position_std=np.sqrt(1.0),
        block_size=16,
        disable_progress_bar=True,
    )

    # Decoding window with no spikes
    t_edges = jnp.linspace(0.0, 10.0, 6)
    empty_spikes = [jnp.array([]), jnp.array([])]
    ll_local = predict_sorted_spikes_kde_log_likelihood(
        time=t_edges,
        position_time=t_pos,
        position=pos,
        spike_times=empty_spikes,
        environment=env,
        marginal_models=enc["marginal_models"],
        occupancy_model=enc["occupancy_model"],
        occupancy=enc["occupancy"],
        mean_rates=jnp.asarray(enc["mean_rates"]),
        place_fields=enc["place_fields"],
        no_spike_part_log_likelihood=enc["no_spike_part_log_likelihood"],
        is_track_interior=enc["is_track_interior"],
        disable_progress_bar=True,
        is_local=True,
    )

    # Compute expected negative sum of local rates at interpolated positions
    interpolated_position = get_position_at_time(t_pos, pos, t_edges, env)
    occupancy_at_time = enc["occupancy_model"].predict(interpolated_position)
    expected = jnp.zeros((t_edges.shape[0],))
    for m, mean_rate in zip(
        enc["marginal_models"], jnp.asarray(enc["mean_rates"]), strict=False
    ):
        marginal = m.predict(interpolated_position)
        marginal = jnp.where(jnp.isnan(marginal), 0.0, marginal)
        local_rate = mean_rate * jnp.where(
            occupancy_at_time > 0.0, marginal / occupancy_at_time, EPS
        )
        local_rate = jnp.clip(local_rate, min=EPS)
        expected -= local_rate
    expected = jnp.expand_dims(expected, axis=1)

    assert jnp.allclose(ll_local, expected, rtol=1e-5, atol=1e-6)


def test_nonlocal_with_no_spikes_equals_negative_no_spike_part(simple_1d_environment):
    env = simple_1d_environment
    t_pos = jnp.linspace(0.0, 10.0, 101)
    pos = jnp.linspace(0.0, 10.0, 101)[:, None]
    spikes_enc = [jnp.array([2.0, 5.0, 5.1]), jnp.array([1.5, 7.2])]
    weights = jnp.ones_like(t_pos)

    enc = fit_sorted_spikes_kde_encoding_model(
        position_time=t_pos,
        position=pos,
        spike_times=spikes_enc,
        environment=env,
        weights=weights,
        sampling_frequency=10,
        position_std=np.sqrt(1.0),
        block_size=16,
        disable_progress_bar=True,
    )

    t_edges = jnp.linspace(0.0, 10.0, 6)
    empty_spikes = [jnp.array([]), jnp.array([])]
    ll = predict_sorted_spikes_kde_log_likelihood(
        time=t_edges,
        position_time=t_pos,
        position=pos,
        spike_times=empty_spikes,
        environment=env,
        marginal_models=enc["marginal_models"],
        occupancy_model=enc["occupancy_model"],
        occupancy=enc["occupancy"],
        mean_rates=jnp.asarray(enc["mean_rates"]),
        place_fields=enc["place_fields"],
        no_spike_part_log_likelihood=enc["no_spike_part_log_likelihood"],
        is_track_interior=enc["is_track_interior"],
        disable_progress_bar=True,
        is_local=False,
    )
    expected = -enc["no_spike_part_log_likelihood"][enc["is_track_interior"]]
    expected = jnp.tile(expected, (t_edges.shape[0], 1))
    assert jnp.allclose(ll, expected, rtol=1e-5, atol=1e-6)


def test_fit_sorted_spikes_kde_raises_without_place_grid():
    env = Environment(environment_name="nofit")
    t_pos = jnp.linspace(0.0, 10.0, 11)
    pos = jnp.linspace(0.0, 10.0, 11)[:, None]
    spikes = [jnp.array([2.0]), jnp.array([3.0])]
    with pytest.raises(ValueError):
        fit_sorted_spikes_kde_encoding_model(
            position_time=t_pos,
            position=pos,
            spike_times=spikes,
            environment=env,
            weights=jnp.ones_like(t_pos),
        )


# ============================================================================
# SNAPSHOT TESTS
# ============================================================================


def serialize_encoding_model_summary(encoding: dict) -> dict:
    """Serialize encoding model to summary statistics for snapshot comparison.

    Parameters
    ----------
    encoding : dict
        Encoding model dictionary from fit_sorted_spikes_kde_encoding_model

    Returns
    -------
    summary : dict
        Summary statistics suitable for snapshot comparison
    """
    return {
        "occupancy_stats": {
            "shape": encoding["occupancy"].shape,
            "mean": float(np.mean(encoding["occupancy"])),
            "std": float(np.std(encoding["occupancy"])),
            "min": float(np.min(encoding["occupancy"])),
            "max": float(np.max(encoding["occupancy"])),
        },
        "mean_rates": [float(r) for r in encoding["mean_rates"]],
        "place_fields_stats": {
            "shape": encoding["place_fields"].shape,
            "mean": float(np.mean(encoding["place_fields"])),
            "std": float(np.std(encoding["place_fields"])),
            "min": float(np.min(encoding["place_fields"])),
            "max": float(np.max(encoding["place_fields"])),
        },
        "no_spike_part_log_likelihood_stats": {
            "shape": encoding["no_spike_part_log_likelihood"].shape,
            "mean": float(np.mean(encoding["no_spike_part_log_likelihood"])),
            "std": float(np.std(encoding["no_spike_part_log_likelihood"])),
            "min": float(np.min(encoding["no_spike_part_log_likelihood"])),
            "max": float(np.max(encoding["no_spike_part_log_likelihood"])),
        },
        "n_neurons": len(encoding["marginal_models"]),
    }


def serialize_log_likelihood_summary(log_likelihood: jnp.ndarray) -> dict:
    """Serialize log likelihood array to summary statistics for snapshot comparison.

    Parameters
    ----------
    log_likelihood : jnp.ndarray
        Log likelihood array from predict_sorted_spikes_kde_log_likelihood

    Returns
    -------
    summary : dict
        Summary statistics suitable for snapshot comparison
    """
    arr = np.asarray(log_likelihood)
    return {
        "shape": arr.shape,
        "dtype": str(arr.dtype),
        "mean": float(np.mean(arr)),
        "std": float(np.std(arr)),
        "min": float(np.min(arr)),
        "max": float(np.max(arr)),
        "sum": float(np.sum(arr)),
        "first_5": arr.ravel()[:5].tolist() if arr.size >= 5 else arr.ravel().tolist(),
        "last_5": arr.ravel()[-5:].tolist() if arr.size >= 5 else arr.ravel().tolist(),
    }


@pytest.mark.snapshot
def test_sorted_spikes_kde_encoding_model_snapshot(
    simple_1d_environment, snapshot: SnapshotAssertion
):
    """Snapshot test for sorted spikes KDE encoding model fitting.

    This test verifies that the encoding model produces consistent outputs
    across code changes, capturing:
    - Occupancy statistics
    - Mean firing rates per neuron
    - Place field statistics
    - No-spike part of log likelihood
    """
    env = simple_1d_environment
    np.random.seed(234)
    t_pos = jnp.linspace(0.0, 10.0, 201)
    pos = jnp.linspace(0.0, 10.0, 201)[:, None]
    weights = jnp.ones_like(t_pos)

    # Four neurons with different spike patterns
    spike_times = [
        jnp.array([1.0, 2.0, 3.0, 5.0, 7.0]),  # 5 spikes
        jnp.array([1.5, 4.0, 6.0, 8.5]),  # 4 spikes
        jnp.array([2.5, 5.5, 7.5, 9.0, 9.5]),  # 5 spikes
        jnp.array([0.5, 3.5, 6.5]),  # 3 spikes
    ]

    encoding = fit_sorted_spikes_kde_encoding_model(
        position_time=t_pos,
        position=pos,
        spike_times=spike_times,
        environment=env,
        weights=weights,
        sampling_frequency=20,
        position_std=np.sqrt(1.5),
        block_size=16,
        disable_progress_bar=True,
    )

    assert serialize_encoding_model_summary(encoding) == snapshot


@pytest.mark.snapshot
def test_sorted_spikes_kde_nonlocal_likelihood_snapshot(
    simple_1d_environment, snapshot: SnapshotAssertion
):
    """Snapshot test for sorted spikes KDE non-local likelihood prediction.

    This test verifies that non-local likelihood predictions are consistent
    across code changes.
    """
    env = simple_1d_environment
    np.random.seed(567)
    t_pos = jnp.linspace(0.0, 10.0, 201)
    pos = jnp.linspace(0.0, 10.0, 201)[:, None]
    weights = jnp.ones_like(t_pos)

    # Two neurons for encoding
    enc_spike_times = [
        jnp.array([2.0, 5.0, 5.1, 7.0]),
        jnp.array([1.5, 3.5, 6.0, 8.0, 9.0]),
    ]

    encoding = fit_sorted_spikes_kde_encoding_model(
        position_time=t_pos,
        position=pos,
        spike_times=enc_spike_times,
        environment=env,
        weights=weights,
        sampling_frequency=20,
        position_std=np.sqrt(1.0),
        block_size=16,
        disable_progress_bar=True,
    )

    # Decoding with different spike times
    t_edges = jnp.linspace(0.0, 10.0, 11)
    dec_spike_times = [
        jnp.array([0.5, 2.5, 5.5, 7.5, 9.5]),
        jnp.array([1.0, 3.0, 6.5, 8.5]),
    ]

    ll_nonlocal = predict_sorted_spikes_kde_log_likelihood(
        time=t_edges,
        position_time=t_pos,
        position=pos,
        spike_times=dec_spike_times,
        environment=env,
        marginal_models=encoding["marginal_models"],
        occupancy_model=encoding["occupancy_model"],
        occupancy=encoding["occupancy"],
        mean_rates=jnp.asarray(encoding["mean_rates"]),
        place_fields=encoding["place_fields"],
        no_spike_part_log_likelihood=encoding["no_spike_part_log_likelihood"],
        is_track_interior=encoding["is_track_interior"],
        disable_progress_bar=True,
        is_local=False,
    )

    assert serialize_log_likelihood_summary(ll_nonlocal) == snapshot


@pytest.mark.snapshot
def test_sorted_spikes_kde_local_likelihood_snapshot(
    simple_1d_environment, snapshot: SnapshotAssertion
):
    """Snapshot test for sorted spikes KDE local likelihood prediction.

    This test verifies that local likelihood predictions are consistent
    across code changes.
    """
    env = simple_1d_environment
    np.random.seed(890)
    t_pos = jnp.linspace(0.0, 10.0, 201)
    pos = jnp.linspace(0.0, 10.0, 201)[:, None]
    weights = jnp.ones_like(t_pos)

    # Three neurons for encoding
    enc_spike_times = [
        jnp.array([2.0, 5.0, 5.1]),
        jnp.array([1.5, 7.2]),
        jnp.array([3.0, 6.0, 8.0]),
    ]

    encoding = fit_sorted_spikes_kde_encoding_model(
        position_time=t_pos,
        position=pos,
        spike_times=enc_spike_times,
        environment=env,
        weights=weights,
        sampling_frequency=20,
        position_std=np.sqrt(1.0),
        block_size=16,
        disable_progress_bar=True,
    )

    # Decoding
    t_edges = jnp.linspace(0.0, 10.0, 6)
    dec_spike_times = [
        jnp.array([2.1, 5.2]),
        jnp.array([1.6]),
        jnp.array([3.1, 6.1, 8.1]),
    ]

    ll_local = predict_sorted_spikes_kde_log_likelihood(
        time=t_edges,
        position_time=t_pos,
        position=pos,
        spike_times=dec_spike_times,
        environment=env,
        marginal_models=encoding["marginal_models"],
        occupancy_model=encoding["occupancy_model"],
        occupancy=encoding["occupancy"],
        mean_rates=jnp.asarray(encoding["mean_rates"]),
        place_fields=encoding["place_fields"],
        no_spike_part_log_likelihood=encoding["no_spike_part_log_likelihood"],
        is_track_interior=encoding["is_track_interior"],
        disable_progress_bar=True,
        is_local=True,
    )

    assert serialize_log_likelihood_summary(ll_local) == snapshot
