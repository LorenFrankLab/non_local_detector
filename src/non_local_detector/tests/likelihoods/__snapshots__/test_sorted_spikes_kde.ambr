# serializer version: 1
# name: test_sorted_spikes_kde_encoding_model_snapshot
  dict({
    'mean_rates': list([
      0.024875622242689133,
      0.019900497049093246,
      0.024875622242689133,
      0.014925372786819935,
    ]),
    'n_neurons': 4,
    'no_spike_part_log_likelihood_stats': dict({
      'max': 0.09434086829423904,
      'mean': 0.08575793355703354,
      'min': 0.07575342804193497,
      'shape': tuple(
        10,
      ),
      'std': 0.005732208024710417,
    }),
    'occupancy_stats': dict({
      'max': 0.09949125349521637,
      'mean': 0.09030207246541977,
      'min': 0.06626125425100327,
      'shape': tuple(
        10,
      ),
      'std': 0.012632736936211586,
    }),
    'place_fields_stats': dict({
      'max': 0.053524088114500046,
      'mean': 0.021439483389258385,
      'min': 0.0012178084580227733,
      'shape': tuple(
        4,
        10,
      ),
      'std': 0.010895122773945332,
    }),
  })
# ---
# name: test_sorted_spikes_kde_local_likelihood_snapshot
  dict({
    'dtype': 'float32',
    'first_5': list([
      -4.384701728820801,
      -8.317499160766602,
      -3.7030904293060303,
      -3.8359668254852295,
      -3.804558753967285,
    ]),
    'last_5': list([
      -8.317499160766602,
      -3.7030904293060303,
      -3.8359668254852295,
      -3.804558753967285,
      -0.006083079148083925,
    ]),
    'max': -0.006083079148083925,
    'mean': -4.0086493492126465,
    'min': -8.317499160766602,
    'shape': tuple(
      6,
      1,
    ),
    'std': 2.411533832550049,
    'sum': -24.051897048950195,
  })
# ---
# name: test_sorted_spikes_kde_nonlocal_likelihood_snapshot
  dict({
    'dtype': 'float32',
    'first_5': list([
      -4.709963798522949,
      -4.01280403137207,
      -3.992805242538452,
      -4.0296101570129395,
      -3.3786098957061768,
    ]),
    'last_5': list([
      -0.06376601755619049,
      -0.05675554275512695,
      -0.050452765077352524,
      -0.04556235671043396,
      -0.035704467445611954,
    ]),
    'max': -0.02684459090232849,
    'mean': -3.2990827560424805,
    'min': -6.716692924499512,
    'shape': tuple(
      11,
      10,
    ),
    'std': 1.69065523147583,
    'sum': -362.89910888671875,
  })
# ---
